### PR TITLE
Player pings

### PR DESCRIPTION
**Pull request by AdamTadeusz**
## Description

Allows players to share a location with teammates

[Discord post](https://discord.com/channels/124171198245502976/1368914502526963843)

[Google Doc](https://docs.google.com/document/d/1TK6s58-cxsZAfJ9X-xEIE1Zwqd5y3URdhbCOnhXF2Ww/edit?usp=sharing)

### CODE REVIEW

**Review by brysondev — COMMENTED** [review]
lgtm; need to test UHD resolutions and networked sounds.

**Comment by brysondev** [comment]
Missing ping material, also scale is a little small on 4K

**Review by brysondev — APPROVED** [review]
image

**Comment by Rainyan** [comment]
@AdamTadeusz What's the current status of this design doc, has this been approved for merge?

**Comment by AdamTadeusz** [comment]
I don't really know who approves these design docs. I think I've addressed the concerns people have as best as I could.

**Review by sunzenshen — APPROVED** [review]
I guess we would need to test this in multiplayer to see how this goes, but this works on my machine in an empty server.  Maybe this could be hidden behind a feature flag if people aren't sure about it.

File: game/neo/resource/NeoModEvents.res [hunk 1]
```diff
@@ -81,6 +81,15 @@
 		"achievement"	"short"		// achievement ID
 	}
 	
+	"player_ping"
+	{
+		"playerindex"	"byte"		// entindex of the player
```
Comment (Rainyan) [inline game/neo/resource/NeoModEvents.res]: We should use userids for events, because we have no time guarantees on that index holding the same client when the event arrives.

File: src/game/client/neo/ui/neo_hud_player_ping.cpp [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+
+#include "iclientmode.h"
+#include <vgui/ILocalize.h>
+#include <vgui/ISurface.h>
+#include "engine/IEngineSound.h"
+#include "voice_status.h"
+#include "hud_chat.h"
+
+#include "ienginevgui.h"
+
```
Comment (Rainyan) [inline src/game/client/neo/ui/neo_hud_player_ping.cpp:17]: Preferably end cpp includes with:
```cpp
// memdbgon must be the last include file in a .cpp file!!!
#include "tier0/memdbgon.h"
```
for consistency

File: src/game/client/neo/ui/neo_hud_player_ping.cpp [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+DECLARE_NAMED_HUDELEMENT(CNEOHud_PlayerPing, NHudPlayerPing);
+
+NEO_HUD_ELEMENT_DECLARE_FREQ_CVAR(PlayerPing, 0.05) // distances to pings and thus ping offset update at 20Hz looks nice enough
+
+CNEOHud_PlayerPing* playerPingHudElement;
+static void clNeoPlayerPingsCallback(IConVar* var, const char* pOldValue, float flOldValue)
```
Comment (Rainyan) [inline src/game/client/neo/ui/neo_hud_player_ping.cpp]: Nit: `[[maybe_unused]]` for `var`, `pOldValue`

File: src/game/client/neo/ui/neo_hud_player_ping.cpp [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+{
+	if (flOldValue && playerPingHudElement)
```
Comment (Rainyan) [inline src/game/client/neo/ui/neo_hud_player_ping.cpp]: Redundant `if (playerPingHudElement)` check, because we declare `CNEOHud_PlayerPing* playerPingHudElement` with no initial value, so it'll likely != nullptr, and we also don't set it to nullptr at a `CNEOHud_PlayerPing` destructor. It might be good to do this stuff just in case, but if we're absolutely confident `playerPingHudElement` will never dangle then the if check could be removed instead.

File: src/game/client/neo/ui/neo_hud_player_ping.cpp [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+	auto *player = C_NEO_Player::GetLocalNEOPlayer();
+	if (!player) return;
+
+	const int playerTeam = player->GetTeamNumber();
+	if (playerTeam != TEAM_JINRAI && playerTeam != TEAM_NSF)
+	{
+		return;
+	}
+
+	for (int i = 0; i < gpGlobals->maxClients; i++)
```
Comment (Rainyan) [inline src/game/client/neo/ui/neo_hud_player_ping.cpp]: Nit: this looks like an off-by-one loop of client indices; could we rename the iterator from `i` to `slot`, so it's more obvious what we're doing here is deliberate for the reader, and not a mistake? The term "player slot" being the Source jargon for "client entindex-1". https://github.com/NeotokyoRebuild/neo/blob/9b5381bf0197ae00f5e9539fe06e1644025b5a60/src/public/iclient.h#L37

Assuming this choice was intentional(?)

File: src/game/client/neo/ui/neo_hud_player_ping.cpp [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+{
+	auto eventName = event->GetName();
+	if (!Q_stricmp(eventName, "player_ping"))
+	{
+		if (!cl_neo_player_pings.GetBool())
+		{
+			return;
+		}
+
+		int playerIndex = event->GetInt("playerindex");
```
Comment (Rainyan) [inline src/game/client/neo/ui/neo_hud_player_ping.cpp]: Rename "player slot" for clarify, since this is (should be?) entindex-1?

File: src/game/shared/neo/neo_player_shared.cpp [hunk 7]
```diff
@@ -112,6 +112,65 @@ bool ClientWantsAimHold(const CNEO_Player* player)
...
+		player->EyeVectors(&forward);
+		Vector eyePosition = player->EyePosition();
+		UTIL_TraceLine(eyePosition, eyePosition + forward * MAX_COORD_RANGE, MASK_VISIBLE_AND_NPCS, player, COLLISION_GROUP_NONE, &tr);
+
+		if (Q_stristr(tr.surface.name, "SKYBOX"))
+		{
+			return;
+		}
+
+		event->SetInt("playerindex", player->entindex());
```
Comment (Rainyan) [inline src/game/shared/neo/neo_player_shared.cpp]: Is this meant to be entindex-1? Since it appears that we're allocating only `MAX_PLAYERS` and not `MAX_PLAYERS_ARRAY_SAFE`? Or are we going off-by-one at some point in the indexing logic?

Anyways, preferably set as "userid", then receiver does userid to clientindex, or clientindex-1 as needed.

File: src/game/client/neo/ui/neo_hud_player_ping.cpp [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+		if (GetClientVoiceMgr()->IsPlayerBlocked(playerIndex))
```
Comment (Rainyan) [inline src/game/client/neo/ui/neo_hud_player_ping.cpp:129]: Off by one for the playerindex \[sic\] call here?

File: src/game/client/neo/ui/neo_hud_player_ping.cpp [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+	}
+
+}
+
+void CNEOHud_PlayerPing::LevelShutdown(void)
+{
+	for (int i = 0; i < MAX_PLAYERS; i++)
+	{
+		m_iPlayerPings[i].deathTime = 0;
+	}
```
Comment (Rainyan) [inline src/game/client/neo/ui/neo_hud_player_ping.cpp]: If we're zeroing the whole range, I'd prefer the loop condition be based on what we're accessing, so `ARRAYSIZE(m_iPlayerPings)`

File: src/game/client/neo/ui/neo_hud_player_ping.cpp [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+
+	return iLength;
+}
+
+void CNEOHud_PlayerPing::HideAllPings()
+{
+	for (int i = 0; i < MAX_PLAYERS; i++)
+	{
+		m_iPlayerPings[i].deathTime = 0;
+	}
```
Comment (Rainyan) [inline src/game/client/neo/ui/neo_hud_player_ping.cpp]: If we're zeroing the whole range, I'd prefer the loop condition be based on what we're accessing, so `ARRAYSIZE(m_iPlayerPings)`. It'll be compile-time evaluated anyway.

File: src/game/client/neo/ui/neo_hud_player_ping.cpp [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+	m_iPlayerPings[pingIndex].distanceYOffset = min(m_iPosY * MAX_Y_DISTANCE_OFFSET_RATIO, m_iPlayerPings[pingIndex].distance * 2 * (m_iPosY / 480));
+
+	trace_t tr;
+	UTIL_TraceLine(player->EyePosition(), m_iPlayerPings[pingIndex].worldPos, MASK_VISIBLE_AND_NPCS, player, COLLISION_GROUP_NONE, &tr);
+	m_iPlayerPings[pingIndex].noLineOfSight = tr.fraction < 0.999;
+}
+
+void CNEOHud_PlayerPing::SetPos(const int index, Vector& pos, bool ghosterPing) {
+	constexpr float PLAYER_PING_LIFETIME = 8;
+	auto localPlayer = UTIL_PlayerByIndex(GetLocalPlayerIndex());
```
Comment (Rainyan) [inline src/game/client/neo/ui/neo_hud_player_ping.cpp]: This could simplify to `C_NEO_Player::GetLocalNEOPlayer()`

File: src/game/client/neo/ui/neo_hud_player_ping.cpp [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+
+	constexpr float MAX_Y_DISTANCE_OFFSET_RATIO = 1.f / 8;
+	m_iPlayerPings[pingIndex].distanceYOffset = min(m_iPosY * MAX_Y_DISTANCE_OFFSET_RATIO, m_iPlayerPings[pingIndex].distance * 2 * (m_iPosY / 480));
+
+	trace_t tr;
+	UTIL_TraceLine(player->EyePosition(), m_iPlayerPings[pingIndex].worldPos, MASK_VISIBLE_AND_NPCS, player, COLLISION_GROUP_NONE, &tr);
+	m_iPlayerPings[pingIndex].noLineOfSight = tr.fraction < 0.999;
+}
+
+void CNEOHud_PlayerPing::SetPos(const int index, Vector& pos, bool ghosterPing) {
```
Comment (Rainyan) [inline src/game/client/neo/ui/neo_hud_player_ping.cpp]: Nit: maybe rename `index` to `playerSlot`, and pass `index-1` in, so it's more clear that we're using player slots. Then `pingIndex` just becomes `playerSlot`.

File: src/game/client/neo/ui/neo_hud_player_ping.h [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+};
+
+class CNEOHud_PlayerPing : public CNEOHud_ChildElement, public CHudElement, public vgui::Panel
+{
+	DECLARE_CLASS_SIMPLE(CNEOHud_PlayerPing, Panel)
+
+public:
+	CNEOHud_PlayerPing(const char *pElementName, vgui::Panel *parent = nullptr);
+
+	void Init(void);
```
Comment (Rainyan) [inline src/game/client/neo/ui/neo_hud_player_ping.h]: Should annotate `virtual void override`

File: src/game/client/neo/ui/neo_hud_player_ping.h [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+	virtual void ApplySchemeSettings(vgui::IScheme *pScheme) override;
+	virtual void Paint() override;
+	virtual void UpdateStateForNeoHudElementDraw() override;
+	virtual void DrawNeoHudElement() override;
+	virtual ConVar *GetUpdateFrequencyConVar() const override;
+	virtual void LevelShutdown(void) override;
+
+	void HideAllPings();
+protected:
+	virtual void FireGameEvent(IGameEvent* event);
```
Comment (Rainyan) [inline src/game/client/neo/ui/neo_hud_player_ping.h]: Preferably annotate `override`

File: src/game/shared/neo/neo_player_shared.cpp [hunk 15]
```diff
@@ -112,6 +112,65 @@ bool ClientWantsAimHold(const CNEO_Player* player)
...
+		event->SetInt("pingx", tr.endpos.x);
+		event->SetInt("pingy", tr.endpos.y);
+		event->SetInt("pingz", tr.endpos.z);
+		event->SetBool("ghosterping", NEORules()->GetGhosterPlayer() == player->entindex() || player->m_iNeoClass == NEO_CLASS_VIP);
```
Comment (Rainyan) [inline src/game/shared/neo/neo_player_shared.cpp]: Might be cleaner to call `player->IsCarryingGhost`

**Review by Rainyan — CHANGES_REQUESTED** [review]

File: src/game/client/neo/ui/neo_hud_player_ping.cpp (hunk 8 above)
Comment (AdamTadeusz) [reply src/game/client/neo/ui/neo_hud_player_ping.cpp:129]: not sure what [sic] means, but if it means that this might be off by one once playerindex is removed from the event, it looks to me like IsPlayerBlocked takes the entity index, not player slot so this should be fine, of course playerindex will have to be grabbed from the player with the matching userid to the one in the event instead

File: src/game/shared/neo/neo_player_shared.cpp (hunk 15 above)
Comment (AdamTadeusz) [reply src/game/shared/neo/neo_player_shared.cpp]: also runs in O(N) instead of constant time, but maybe we should use IsCarryingGhost anyway and maybe refactor how weapons are stored somewhere down the line to not use an array that weapons can fall into in any order

File: src/game/client/neo/ui/neo_hud_player_ping.cpp (hunk 11 above)
Comment (AdamTadeusz) [reply src/game/client/neo/ui/neo_hud_player_ping.cpp]: localPlayer doesn't get cast to C_NEO_Player anywhere, unless you're saying it should anyway.

If the comment is for the player pointer cast further down, I don't see a shorthand for casting a player at a specific index to a C_NEO_Player (I can add one?)

File: src/game/shared/neo/neo_player_shared.cpp (hunk 15 above)
Comment (Rainyan) [reply src/game/shared/neo/neo_player_shared.cpp]: Yeah it's indeed faster, but like you said I think it'd be better to optimize the helper method itself later, so we could go for code readability. But I'm ok with whichever you think fits here better. Should've marked this as a nit.

File: src/game/client/neo/ui/neo_hud_player_ping.cpp (hunk 11 above)
Comment (Rainyan) [reply src/game/client/neo/ui/neo_hud_player_ping.cpp]: > localPlayer doesn't get cast to C_NEO_Player anywhere, unless you're saying it should anyway.

I meant that we could do:
```diff
index b65a8671..fe6de30a 100644
--- a/src/game/client/neo/ui/neo_hud_player_ping.cpp
+++ b/src/game/client/neo/ui/neo_hud_player_ping.cpp
@@ -287,7 +287,7 @@ void CNEOHud_PlayerPing::UpdateDistanceToPlayer(C_BasePlayer* player, int pingIn

 void CNEOHud_PlayerPing::SetPos(const int index, Vector& pos, bool ghosterPing) {
        constexpr float PLAYER_PING_LIFETIME = 8;
-       auto localPlayer = UTIL_PlayerByIndex(GetLocalPlayerIndex());
+       auto localPlayer = C_NEO_Player::GetLocalNEOPlayer();
        if (!localPlayer) { return; }
        auto pingPlayer = static_cast<C_NEO_Player*>(UTIL_PlayerByIndex(index));
        if (!pingPlayer) { return; }
```
to avoid the loop ran by `UTIL_PlayerByIndex` for getting the local player.

Or `C_BasePlayer::GetLocalPlayer()`, but it's gonna return the same pointer value, just as a different pointer type.

File: src/game/client/neo/ui/neo_hud_player_ping.cpp (hunk 8 above)
Comment (Rainyan) [reply src/game/client/neo/ui/neo_hud_player_ping.cpp:129]: I can't quite remember the context anymore, but this was related to my confusion over whether we are zero-indexing or one-indexing the values being passed around by the event. But in either case, we should probably prefer userids, as the player being represented by a slot/index might theoretically be discarted/reused before the listener receives the event.

Comment (Rainyan) [reply src/game/client/neo/ui/neo_hud_player_ping.cpp:129]: I think my main confusion was this: https://github.com/NeotokyoRebuild/neo/pull/1125#discussion_r2247868845

So we're allocating space for zero-indexing the array, but entindices are one-indexed, so it'd appear like we're going off-by-one at some point in the logic. But I can do a re-review if you think the logic is correct as it is, I could have been mistaken when looking at it previously.

Comment (AdamTadeusz) [reply src/game/client/neo/ui/neo_hud_player_ping.cpp:129]: lemme work through all the comments again and understand it in my head and push a commit

File: src/game/client/neo/ui/neo_hud_player_ping.cpp (hunk 5 above)
Comment (Rainyan) [reply src/game/client/neo/ui/neo_hud_player_ping.cpp]: Ok, this looks like I was wrong, since indeed we are zero-indexing the "ping index", and not "client index" in the subsequent call. So I think this is actually working correctly already(?)

File: src/game/client/neo/ui/neo_hud_player_ping.cpp (hunk 8 above)
Comment (Rainyan) [reply src/game/client/neo/ui/neo_hud_player_ping.cpp:129]: I might have been confused about "player index" vs "ping index" in some of the code. Will take another look later today with fresh eyes.

File: src/game/shared/neo/neo_player_shared.cpp [hunk 16]
```diff
@@ -112,6 +112,65 @@ bool ClientWantsAimHold(const CNEO_Player* player)
...
+#endif // GAME_DLL
+		constexpr float NEO_PING_DELAY = 2.f;
+		if ((gpGlobals->curtime - player->m_flNextPingTime) < NEO_PING_DELAY)
+		{ // NEO TODO (Adam) fix for pings placed during the first NEO_PING_DELAY seconds of the server's life?
+			player->m_flNextPingTime = gpGlobals->curtime + NEO_PING_DELAY;
+		}
+		else
+		{
+			player->m_flNextPingTime = gpGlobals->curtime;
+		}
```
Comment (Rainyan) [inline src/game/shared/neo/neo_player_shared.cpp:167]: There appears to be a problem here where the first two pings can be placed rapidly. Could this be simplified to:
```diff
index 720d1540..4511a032 100644
--- a/src/game/shared/neo/neo_player_shared.cpp
+++ b/src/game/shared/neo/neo_player_shared.cpp
@@ -160,14 +160,7 @@ void CheckPingButton(CNEO_Player* player)
                gameeventmanager->FireEventClientSide(event);
 #endif // GAME_DLL
                constexpr float NEO_PING_DELAY = 2.f;
-               if ((gpGlobals->curtime - player->m_flNextPingTime) < NEO_PING_DELAY)
-               { // NEO TODO (Adam) fix for pings placed during the first NEO_PING_DELAY seconds of the server's life?
-                       player->m_flNextPingTime = gpGlobals->curtime + NEO_PING_DELAY;
-               }
-               else
-               {
-                       player->m_flNextPingTime = gpGlobals->curtime;
-               }
+               player->m_flNextPingTime = gpGlobals->curtime + NEO_PING_DELAY;
        }
 }
```
since I don't think there's anything preventing us from unconditionally setting the next ping time as curtime+delay?

**Review by Rainyan — CHANGES_REQUESTED** [review]

File: src/game/shared/neo/neo_player_shared.cpp (hunk 16 above)
Comment (AdamTadeusz) [reply src/game/shared/neo/neo_player_shared.cpp:167]: this is intentional. Pings get placed on mouse press so there's little chance of a player accidentally placing a ping twice in quick succession unless they want to adjust where their ping was placed.

File: src/game/client/neo/ui/neo_hud_player_ping.cpp [hunk 17]
```diff
@@ -0,0 +1,349 @@
...
+{
+	if (cl_neo_player_pings_chat_message.GetBool() && pPlayer && !pPlayer->IsLocalPlayer())
+	{
+		CBaseHudChat* hudChat = (CBaseHudChat*)GET_HUDELEMENT(CHudChat);
+		if (hudChat)
+		{
+			char szText[256];
+			V_strcpy_safe(szText, pPlayer->GetNeoPlayerName());
+			V_strcat_safe(szText, " pinged a location\n");
+			hudChat->ChatPrintf(0, CHAT_FILTER_NONE, szText);
```
Comment (Rainyan) [inline src/game/client/neo/ui/neo_hud_player_ping.cpp:373]: This message is printed to the enemy when their opponent uses a ping. That should probably not happen.

File: src/game/client/neo/ui/neo_hud_player_ping.cpp [hunk 18]
```diff
@@ -0,0 +1,349 @@
...
+	if (pingSoundHandle == -1)
+	{
+		et.m_pSoundName = "gameplay/ping.wav";
+	}
+	else
+	{
+		et.m_hSoundScriptHandle = pingSoundHandle;
+	}
+	et.m_nFlags |= SND_CHANGE_VOL;
+	et.m_flVolume = snd_ping_volume.GetFloat();
```
Comment (Rainyan) [inline src/game/client/neo/ui/neo_hud_player_ping.cpp]: It might be nice to set the ping origin for the `EmitSound_t` to the ping location, so you could estimate where the ping is placed, even if it happens outside your screen.

File: src/game/client/neo/ui/neo_hud_player_ping.cpp (hunk 17 above)
Comment (AdamTadeusz) [reply src/game/client/neo/ui/neo_hud_player_ping.cpp:373]: looks like UTIL_PlayerByUserId only returns players if the client knows about them, so if they are or previously were in pvs. I guess I should use some other function to get the player index from the user ID.

Comment (AdamTadeusz) [reply src/game/client/neo/ui/neo_hud_player_ping.cpp:373]: this gives me an idea for how improve the spectator xray, I can probably add something to NotifyShouldTransmit which gets triggered when a player begins being transmitted to the local player, to update their glow state and glow colour as I should at that point have their team and an object which I can add the glow effect to

Comment (Rainyan) [reply src/game/client/neo/ui/neo_hud_player_ping.cpp:373]: > looks like UTIL_PlayerByUserId only returns players if the client knows about them, so if they are or previously were in pvs. I guess I should use some other function to get the player index from the user ID.

One alternative could be to split the ping event into 2 events, one for each team. And then the players would only listen for the event that corresponds to their current team. Then you didn't need to do the team check when receiving the event, and could for example pass the pinger username in the event directly so the client didn't need to do the userid lookup at all.

Comment (AdamTadeusz) [reply src/game/client/neo/ui/neo_hud_player_ping.cpp:373]: would you rather I did that or passed the team as another parameter in the event?

File: src/game/client/neo/ui/neo_hud_player_ping.cpp (hunk 18 above)
Comment (AdamTadeusz) [reply src/game/client/neo/ui/neo_hud_player_ping.cpp]: the sound origin takes a pointer to a const Vector, not sure how to go about passing that in since I need to be able to modify the origin in the pointers list, and a variable the value is copied to may go out of scope while the pointer is still being used, I guess I should look at how its used elsewhere

File: src/game/client/neo/ui/neo_hud_player_ping.cpp (hunk 17 above)
Comment (Rainyan) [reply src/game/client/neo/ui/neo_hud_player_ping.cpp:373]: > would you rather I did that or passed the team as another parameter in the event?

Either way works for me, your suggestion might be cleaner though so maybe go with that?

File: src/game/client/neo/ui/neo_hud_player_ping.cpp (hunk 18 above)
Comment (Rainyan) [reply src/game/client/neo/ui/neo_hud_player_ping.cpp]: > the sound origin takes a pointer to a const Vector, not sure how to go about passing that in since I need to be able to modify the origin in the pointers list, and a variable the value is copied to may go out of scope while the pointer is still being used, I guess I should look at how its used elsewhere

Ok, we can leave it outside this PR and maybe revisit later. No need to make things overly complicated for now.

File: src/game/shared/neo/neo_player_shared.cpp [hunk 19]
```diff
@@ -110,6 +110,66 @@ bool ClientWantsAimHold(const CNEO_Player* player)
...
+		Vector eyePosition = player->EyePosition();
+		UTIL_TraceLine(eyePosition, eyePosition + forward * MAX_COORD_RANGE, MASK_VISIBLE_AND_NPCS, player, COLLISION_GROUP_NONE, &tr);
+
+		if (Q_stristr(tr.surface.name, "SKYBOX"))
+		{
+			return;
+		}
+
+		event->SetInt("userid", player->GetUserID());
+		event->SetInt("playerTeam", player->GetTeamNumber());
```
Comment (Rainyan) [inline src/game/shared/neo/neo_player_shared.cpp]: We're setting `playerTeam` here, but reading `playerteam` (different capitalization) in the `CNEOHud_PlayerPing::FireGameEvent`. Since the event code is owned by the engine and we don't have control over the implementation, maybe it'd be safer to unify the capitalization here, just in case so that we can't spontaneously break if Valve ships an update to the engine that was incompatible with case-insensitive logic.

**Review by Rainyan — CHANGES_REQUESTED** [review]

**Review by sunzenshen — APPROVED** [review]

**Review by sunzenshen — APPROVED** [review]
ping_okay
